rclcpp/wait_set/thread_safe_wait_set.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <memory>
+#include <vector>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+int main(int argc, char * argv[])
```

Comment:
This example is showing off the dynamic nature of the wait set, but not really anything about the thread-safe part. Maybe we should remove, or rewrite, these examples that are more like tests.

Reply:
I think it is worth to have one example for the thread safe wait-set at least to make it more visible for the users. We could remove all the dynamic add/remove stuff and simplify it just to show how it is used. Another option would be to create a more specific example, like sharing data between threads. It won't be too impressive probably but it could be a good minimal example where it makes sense to use it.

Reply:
I was thinking more like an example that initially adds some things, runs the wait set in a thread, then adds/removes items concurrently, to show that it's safe.

I mean, "using" the thread-safe wait set is really just creating one and doing everything else normally. Demonstrating or proving that it works is harder to do. I think, as trivial as it is, just having the minimal example that creates a thread-safe wait set instead of a normal is just fine.